File: server/src/modules/delivery/person-archive.service.ts
import { sql, eq, and, desc, isNull } from 'drizzle-orm';
import { db } from '../../config/database.js';
import {
  users,
  careEpisodes,
  groupInstances,
  groupEnrollments,
  courseInstances,
  courseEnrollments,
  assessments,
  assessmentResults,
} from '../../db/schema.js';
import { NotFoundError } from '../../lib/errors.js';

/**
 * Phase 6 — Person archive aggregation (server-side).
 *
 * "Person archive" = the complete cross-module service history of a single
 * user (来访者 / 学员 / 受测者). Same person, four kinds of touchpoints:
 *   - care_episodes WHERE client_id = userId
 *   - group_enrollments WHERE user_id = userId → JOIN group_instances
 *   - course_enrollments WHERE user_id = userId → JOIN course_instances
 *   - assessment_results WHERE user_id = userId → JOIN assessments
 *
 * Two endpoints in this module:
 *   listPeople(orgId)          → who has any service touchpoint, sorted by
 *                                 most-recent activity. For PeopleList.tsx.
 *   getPersonArchive(orgId, userId) → full archive of a single user. For
 *                                 PersonArchive.tsx.
 *
 * Status mapping mirrors `delivery.service.ts` (Phase 5b). Keep them in sync.
 */

export interface PersonSummary {
  userId: string;
  name: string;
  email: string | null;
  /** ISO timestamp of the most recent activity across all 4 services */
  lastActivityAt: string;
  /** Per-kind touchpoint counts */
  counts: {
    counseling: number;
    group: number;
    course: number;
    assessment: number;
    total: number;
  };
}

export interface PersonArchive {
  user: {
    id: string;
    name: string;
    email: string | null;
    avatarUrl: string | null;
  };
  stats: {
    counseling: number;
    group: number;
    course: number;
    assessment: number;
    total: number;
  };
  /**
   * Unified service list (matches the ServiceInstance shape used by
   * DeliveryCard). Includes all services this user has participated in.
   */
  services: ArchivedService[];
  /** Chronological event timeline (oldest → newest) */
  timeline: ArchiveTimelineEvent[];
}

export interface ArchivedService {
  id: string;
  kind: 'counseling' | 'group' | 'course' | 'assessment';
  orgId: string;
  title: string;
  status: string;
  description: string | null;
  /** When this user joined / was first associated with the service */
  joinedAt: string | null;
  /** When the service itself was last updated */
  lastActivityAt: string;
  /** Optional: course/group instance id if relevant for deep linking */
  instanceId: string | null;
  /** counseling-only: chief complaint */
  chiefComplaint: string | null;
  /** counseling-only: current risk */
  currentRisk: string | null;
  /** assessment-only: total score */
  totalScore: number | null;
}

export interface ArchiveTimelineEvent {
  id: string;
  kind: 'counseling' | 'group' | 'course' | 'assessment';
  type: 'episode_opened' | 'episode_closed' | 'group_enrolled' | 'course_enrolled' | 'assessment_taken';
  at: string;
  title: string;
  detail?: string;
  /** Reference to the underlying service */
  serviceId: string;
}

// ─── List people ─────────────────────────────────────────────────

/**
 * Return all users in this org who are either:
 *   - members with role='client' (freshly registered, even before any touchpoint)
 *   - have at least one service touchpoint (counseling / group / course / assessment)
 *
 * Implementation: SQL UNION across 4 touchpoint tables PLUS the org_members table
 * (filtered to role='client'), grouped by user_id with LEFT JOIN to `users` for
 * name/email. The membership branch ensures freshly registered C-side users
 * (no service activity yet) are visible to counselors so they can be assigned
 * to episodes / appointments. Membership rows yield kind='member' which
 * doesn't count toward any per-kind counter (counseling/group/course/assessment
 * remain 0 until a real touchpoint exists), but the user shows up in the list.
 */
export async function listPeople(orgId: string, limit = 200): Promise<PersonSummary[]> {
  const cap = Math.min(Math.max(limit, 1), 1000);

  const result = await db.execute(sql`
    WITH touchpoints AS (
      SELECT
        ce.client_id::text  AS user_id,
        'counseling'::text  AS kind,
        ce.updated_at       AS last_activity_at
      FROM care_episodes ce
      WHERE ce.org_id = ${orgId}

      UNION ALL

      SELECT
        ge.user_id::text    AS user_id,
        'group'::text       AS kind,
        COALESCE(ge.enrolled_at, ge.created_at, gi.updated_at) AS last_activity_at
      FROM group_enrollments ge
      INNER JOIN group_instances gi ON gi.id = ge.instance_id
      WHERE gi.org_id = ${orgId}

      UNION ALL

      SELECT
        cen.user_id::text   AS user_id,
        'course'::text      AS kind,
        COALESCE(cen.enrolled_at, ci.updated_at) AS last_activity_at
      FROM course_enrollments cen
      INNER JOIN course_instances ci ON ci.id = cen.instance_id
      WHERE ci.org_id = ${orgId}

      UNION ALL

      SELECT
        ar.user_id::text    AS user_id,
        'assessment'::text  AS kind,
        ar.created_at       AS last_activity_at
      FROM assessment_results ar
      WHERE ar.org_id = ${orgId}
        AND ar.user_id IS NOT NULL
        AND ar.deleted_at IS NULL

      UNION ALL

      -- Bare membership: client members with no touchpoint yet still appear
      -- so counselors can find freshly-registered users to schedule with.
      SELECT
        om.user_id::text  AS user_id,
        'member'::text    AS kind,
        COALESCE(om.created_at, NOW()) AS last_activity_at
      FROM org_members om
      WHERE om.org_id = ${orgId}
        AND om.role = 'client'
        AND om.status = 'active'
    )
    SELECT
      t.user_id,
      u.name,
      u.email,
      MAX(t.last_activity_at) AS last_activity_at,
      COUNT(*) FILTER (WHERE t.kind = 'counseling') AS counseling,
      COUNT(*) FILTER (WHERE t.kind = 'group')      AS group_count,
      COUNT(*) FILTER (WHERE t.kind = 'course')     AS course_count,
      COUNT(*) FILTER (WHERE t.kind = 'assessment') AS assessment
    FROM touchpoints t
    LEFT JOIN users u ON u.id = t.user_id::uuid
    GROUP BY t.user_id, u.name, u.email
    ORDER BY MAX(t.last_activity_at) DESC NULLS LAST
    LIMIT ${cap}
  `);

  const rawRows: any[] = Array.isArray(result)
    ? (result as any[])
    : ((result as any).rows ?? []);

  return rawRows.map((row): PersonSummary => {
    const counseling = Number(row.counseling ?? 0);
    const group = Number(row.group_count ?? 0);
    const course = Number(row.course_count ?? 0);
    const assessment = Number(row.assessment ?? 0);
    return {
      userId: String(row.user_id),
      name: row.name ? String(row.name) : '未知用户',
      email: row.email ? String(row.email) : null,
      lastActivityAt: new Date(row.last_activity_at).toISOString(),
      counts: {
        counseling,
        group,
        course,
        assessment,
        total: counseling + group + course + assessment,
      },
    };
  });
}

// ─── Get one person's full archive ───────────────────────────────

const STATUS_MAP_EPISODE: Record<string, string> = {
  active: 'ongoing',
  paused: 'paused',
  closed: 'closed',
  archived: 'archived',
};
const STATUS_MAP_GROUP: Record<string, string> = {
  draft: 'draft',
  recruiting: 'recruiting',
  ongoing: 'ongoing',
  full: 'ongoing',
  ended: 'completed',
};
const STATUS_MAP_COURSE: Record<string, string> = {
  draft: 'draft',
  active: 'ongoing',
  closed: 'closed',
  archived: 'archived',
};

function mapAssessmentStatus(status: string, isActive: boolean): string {
  if (status === 'draft') return 'draft';
  if (status === 'archived') return 'archived';
  return isActive ? 'ongoing' : 'paused';
}

/**
 * Fetch the complete archive for a single user. Uses 4 parallel queries
 * (one per kind) plus 1 query for the user record. The whole call is ~5
 * round-trips to PG, all parallelizable, so latency ≈ slowest single query.
 *
 * Throws NotFoundError if the user doesn't exist (regardless of whether they
 * have any touchpoints — non-existent users get a hard 404).
 */
export async function getPersonArchive(orgId: string, userId: string): Promise<PersonArchive> {
  // Run user lookup and 4 service queries in parallel.
  const [userRow, episodesRows, groupRows, courseRows, assessmentRows] = await Promise.all([
    db.select().from(users).where(eq(users.id, userId)).limit(1),
    db
      .select()
      .from(careEpisodes)
      .where(and(eq(careEpisodes.orgId, orgId), eq(careEpisodes.clientId, userId)))
      .orderBy(desc(careEpisodes.updatedAt)),
    db
      .select({
        enrollment: groupEnrollments,
        instance: groupInstances,
      })
      .from(groupEnrollments)
      .innerJoin(groupInstances, eq(groupInstances.id, groupEnrollments.instanceId))
      .where(and(eq(groupEnrollments.userId, userId), eq(groupInstances.orgId, orgId)))
      .orderBy(desc(groupInstances.updatedAt)),
    db
      .select({
        enrollment: courseEnrollments,
        instance: courseInstances,
      })
      .from(courseEnrollments)
      .innerJoin(courseInstances, eq(courseInstances.id, courseEnrollments.instanceId))
      .where(and(eq(courseEnrollments.userId, userId), eq(courseInstances.orgId, orgId)))
      .orderBy(desc(courseInstances.updatedAt)),
    db
      .select({
        result: assessmentResults,
        assessment: assessments,
      })
      .from(assessmentResults)
      .innerJoin(assessments, eq(assessments.id, assessmentResults.assessmentId))
      .where(
        and(
          eq(assessmentResults.orgId, orgId),
          eq(assessmentResults.userId, userId),
          isNull(assessmentResults.deletedAt),
        ),
      )
      .orderBy(desc(assessmentResults.createdAt)),
  ]);

  if (userRow.length === 0) {
    throw new NotFoundError('User', userId);
  }
  const u = userRow[0];

  // Build the unified service list. Each kind contributes its own row shape
  // mapped onto ArchivedService.
  const services: ArchivedService[] = [];
  const timeline: ArchiveTimelineEvent[] = [];

  for (const e of episodesRows) {
    services.push({
      id: e.id,
      kind: 'counseling',
      orgId: e.orgId,
      title: u.name || '未知来访者',
      status: STATUS_MAP_EPISODE[e.status] ?? 'draft',
      description: e.chiefComplaint,
      joinedAt: e.openedAt ? new Date(e.openedAt).toISOString() : null,
      lastActivityAt: new Date(e.updatedAt).toISOString(),
      instanceId: e.id,
      chiefComplaint: e.chiefComplaint,
      currentRisk: e.currentRisk,
      totalScore: null,
    });
    timeline.push({
      id: `ep-open-${e.id}`,
      kind: 'counseling',
      type: 'episode_opened',
      at: new Date(e.openedAt).toISOString(),
      title: '建立个案',
      detail: e.chiefComplaint || undefined,
      serviceId: e.id,
    });
    if (e.closedAt) {
      timeline.push({
        id: `ep-close-${e.id}`,
        kind: 'counseling',
        type: 'episode_closed',
        at: new Date(e.closedAt).toISOString(),
        title: '个案结案',
        serviceId: e.id,
      });
    }
  }

  for (const row of groupRows) {
    const inst = row.instance;
    const enr = row.enrollment;
    services.push({
      id: inst.id,
      kind: 'group',
      orgId: inst.orgId,
      title: inst.title,
      status: STATUS_MAP_GROUP[inst.status] ?? 'draft',
      description: inst.description,
      joinedAt: enr.enrolledAt
        ? new Date(enr.enrolledAt).toISOString()
        : enr.createdAt
          ? new Date(enr.createdAt).toISOString()
          : null,
      lastActivityAt: new Date(inst.updatedAt).toISOString(),
      instanceId: inst.id,
      chiefComplaint: null,
      currentRisk: null,
      totalScore: null,
    });
    if (enr.enrolledAt || enr.createdAt) {
      timeline.push({
        id: `grp-${enr.id}`,
        kind: 'group',
        type: 'group_enrolled',
        at: new Date((enr.enrolledAt ?? enr.createdAt) as Date | string).toISOString(),
        title: `加入团辅: ${inst.title}`,
        serviceId: inst.id,
      });
    }
  }

  for (const row of courseRows) {
    const inst = row.instance;
    const enr = row.enrollment;
    services.push({
      id: inst.id,
      kind: 'course',
      orgId: inst.orgId,
      title: inst.title,
      status: STATUS_MAP_COURSE[inst.status] ?? 'draft',
      description: inst.description,
      joinedAt: enr.enrolledAt ? new Date(enr.enrolledAt).toISOString() : null,
      lastActivityAt: new Date(inst.updatedAt).toISOString(),
      instanceId: inst.id,
      chiefComplaint: null,
      currentRisk: null,
      totalScore: null,
    });
    if (enr.enrolledAt) {
      timeline.push({
        id: `crs-${enr.id}`,
        kind: 'course',
        type: 'course_enrolled',
        at: new Date(enr.enrolledAt).toISOString(),
        title: `加入课程: ${inst.title}`,
        serviceId: inst.id,
      });
    }
  }

  for (const row of assessmentRows) {
    const a = row.assessment;
    const r = row.result;
    services.push({
      id: a.id,
      kind: 'assessment',
      orgId: a.orgId,
      title: a.title,
      status: mapAssessmentStatus(a.status, a.isActive),
      description: a.description,
      joinedAt: r.createdAt ? new Date(r.createdAt).toISOString() : null,
      lastActivityAt: new Date(a.updatedAt).toISOString(),
      instanceId: r.id,
      chiefComplaint: null,
      currentRisk: null,
      totalScore: r.totalScore !== null && r.totalScore !== undefined ? Number(r.totalScore) : null,
    });
    timeline.push({
      id: `asm-${r.id}`,
      kind: 'assessment',
      type: 'assessment_taken',
      at: new Date(r.createdAt).toISOString(),
      title: `完成测评: ${a.title}`,
      detail: r.totalScore !== null ? `总分 ${r.totalScore}` : undefined,
      serviceId: a.id,
    });
  }

  // Deduplicate services by (kind, id) — a single user can have multiple
  // assessment results for the same assessment. We keep the most recent
  // joined-at as the "joinedAt" for the dedupe entry.
  const seen = new Map<string, ArchivedService>();
  for (const s of services) {
    const key = `${s.kind}-${s.id}`;
    const prev = seen.get(key);
    if (!prev) {
      seen.set(key, s);
    } else {
      // Prefer the more recent joinedAt
      if (s.joinedAt && (!prev.joinedAt || s.joinedAt > prev.joinedAt)) {
        prev.joinedAt = s.joinedAt;
      }
      // If this row has a totalScore and prev doesn't, take it
      if (s.totalScore !== null && prev.totalScore === null) {
        prev.totalScore = s.totalScore;
      }
    }
  }
  const uniqueServices = Array.from(seen.values()).sort((a, b) =>
    b.lastActivityAt.localeCompare(a.lastActivityAt),
  );

  // Sort timeline ascending (oldest → newest)
  timeline.sort((a, b) => a.at.localeCompare(b.at));

  const stats = {
    counseling: episodesRows.length,
    group: groupRows.length,
    course: courseRows.length,
    assessment: assessmentRows.length,
    total: episodesRows.length + groupRows.length + courseRows.length + assessmentRows.length,
  };

  return {
    user: {
      id: u.id,
      name: u.name,
      email: u.email,
      avatarUrl: u.avatarUrl,
    },
    stats,
    services: uniqueServices,
    timeline,
  };
}

File: server-py/app/api/v1/delivery/person_archive_service.py
"""
Person archive 服务层 — 镜像 ``server/src/modules/delivery/person-archive.service.ts`` (479 行)。

"Person archive" = 单一人员跨 4 模块的全部 service touchpoint:
  - care_episodes  WHERE client_id = userId
  - group_enrollments → JOIN group_instances WHERE org_id
  - course_enrollments → JOIN course_instances WHERE org_id
  - assessment_results → JOIN assessments WHERE org_id

两个 entry:
  - ``list_people(db, org_id, limit)``       — 列表 + 计数 (PeopleList.tsx)
  - ``get_person_archive(db, org_id, user_id)`` — 单人完整档案 (PersonArchive.tsx)

跨表大聚合, N+1 风险高, 实现策略 (与 Node 一致):

  list_people: **单 raw SQL UNION ALL** (5 分支: 4 个 touchpoint + bare membership),
               外层 GROUP BY 一次出全部用户 + 最后活动时间 + 各 kind 计数。
               比 fan-out 4 query + 客户端聚合性能高一个数量级。

  get_person_archive: **5 个 SQLAlchemy ``select`` 并行** (asyncio.gather) —
               user lookup + 4 个 service touchpoint 查询并发跑 (5 round-trip
               时间 ≈ 最慢一条), 不是串行。每条都按 (orgId + userId) 索引覆盖,
               不会 full scan。

状态映射 (STATUS_MAP_*) 与 ``delivery.service.ts`` (Phase 5b) 同步, 两边一改两边改。
"""

from __future__ import annotations

import asyncio
from typing import Any

from sqlalchemy import and_, desc, select, text
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.v1.delivery.schemas import (
    ArchivedService,
    ArchiveTimelineEvent,
    ArchiveTimelineEventType,
    ListPeopleResponse,
    PersonArchive,
    PersonArchiveUser,
    PersonCounts,
    PersonSummary,
)
from app.db.models.assessment_results import AssessmentResult
from app.db.models.assessments import Assessment
from app.db.models.care_episodes import CareEpisode
from app.db.models.course_enrollments import CourseEnrollment
from app.db.models.course_instances import CourseInstance
from app.db.models.group_enrollments import GroupEnrollment
from app.db.models.group_instances import GroupInstance
from app.db.models.users import User
from app.lib.errors import NotFoundError
from app.lib.uuid_utils import parse_uuid_or_raise

# ─── 状态映射 (镜像 person-archive.service.ts:219-243) ──────────


_STATUS_MAP_EPISODE: dict[str, str] = {
    "active": "ongoing",
    "paused": "paused",
    "closed": "closed",
    "archived": "archived",
}
_STATUS_MAP_GROUP: dict[str, str] = {
    "draft": "draft",
    "recruiting": "recruiting",
    "ongoing": "ongoing",
    "full": "ongoing",
    "ended": "completed",
}
_STATUS_MAP_COURSE: dict[str, str] = {
    "draft": "draft",
    "active": "ongoing",
    "closed": "closed",
    "archived": "archived",
}


def _map_assessment_status(status: str | None, is_active: bool | None) -> str:
    """assessment 状态分支 (镜像 person-archive.service.ts:239-243)。"""
    if status == "draft":
        return "draft"
    if status == "archived":
        return "archived"
    return "ongoing" if is_active else "paused"


def _iso(value: Any) -> str | None:
    """datetime / date / None → ISO str 或 None (集中处理 wire 一致性)。"""
    if value is None:
        return None
    if isinstance(value, str):
        return value
    return value.isoformat() if hasattr(value, "isoformat") else str(value)


# ─── list_people (镜像 person-archive.service.ts:120-215) ─────


async def list_people(
    db: AsyncSession,
    org_id: str,
    limit: int | None = 200,
) -> ListPeopleResponse:
    """列出该 org 内所有有 service touchpoint OR 仅成员身份的客户。

    实现: 单 SQL CTE + UNION ALL 5 分支 (4 touchpoint + bare membership).
    membership 分支保证刚注册无 touchpoint 的 client 仍能被咨询师在列表里找到。
    """
    parse_uuid_or_raise(org_id, field="orgId")
    cap = min(max(limit if limit is not None else 200, 1), 1000)

    sql = text(
        """
        WITH touchpoints AS (
            SELECT
                ce.client_id::text  AS user_id,
                'counseling'::text  AS kind,
                ce.updated_at       AS last_activity_at
            FROM care_episodes ce
            WHERE ce.org_id = :org_id

            UNION ALL

            SELECT
                ge.user_id::text    AS user_id,
                'group'::text       AS kind,
                COALESCE(ge.enrolled_at, ge.created_at, gi.updated_at) AS last_activity_at
            FROM group_enrollments ge
            INNER JOIN group_instances gi ON gi.id = ge.instance_id
            WHERE gi.org_id = :org_id

            UNION ALL

            SELECT
                cen.user_id::text   AS user_id,
                'course'::text      AS kind,
                COALESCE(cen.enrolled_at, ci.updated_at) AS last_activity_at
            FROM course_enrollments cen
            INNER JOIN course_instances ci ON ci.id = cen.instance_id
            WHERE ci.org_id = :org_id

            UNION ALL

            SELECT
                ar.user_id::text    AS user_id,
                'assessment'::text  AS kind,
                ar.created_at       AS last_activity_at
            FROM assessment_results ar
            WHERE ar.org_id = :org_id
              AND ar.user_id IS NOT NULL
              AND ar.deleted_at IS NULL

            UNION ALL

            -- bare membership: 仅 active client member, 即使无 touchpoint
            SELECT
                om.user_id::text  AS user_id,
                'member'::text    AS kind,
                COALESCE(om.created_at, NOW()) AS last_activity_at
            FROM org_members om
            WHERE om.org_id = :org_id
              AND om.role = 'client'
              AND om.status = 'active'
        )
        SELECT
            t.user_id,
            u.name,
            u.email,
            MAX(t.last_activity_at) AS last_activity_at,
            COUNT(*) FILTER (WHERE t.kind = 'counseling') AS counseling,
            COUNT(*) FILTER (WHERE t.kind = 'group')      AS group_count,
            COUNT(*) FILTER (WHERE t.kind = 'course')     AS course_count,
            COUNT(*) FILTER (WHERE t.kind = 'assessment') AS assessment
        FROM touchpoints t
        LEFT JOIN users u ON u.id = t.user_id::uuid
        GROUP BY t.user_id, u.name, u.email
        ORDER BY MAX(t.last_activity_at) DESC NULLS LAST
        LIMIT :cap
        """
    )

    result = await db.execute(sql, {"org_id": org_id, "cap": cap})
    rows = list(result.mappings().all())

    items: list[PersonSummary] = []
    for row in rows:
        counseling = int(row["counseling"] or 0)
        group = int(row["group_count"] or 0)
        course = int(row["course_count"] or 0)
        assessment = int(row["assessment"] or 0)
        items.append(
            PersonSummary(
                user_id=str(row["user_id"]),
                name=str(row["name"]) if row.get("name") else "未知用户",
                email=str(row["email"]) if row.get("email") else None,
                last_activity_at=_iso(row["last_activity_at"]) or "",
                counts=PersonCounts(
                    counseling=counseling,
                    group=group,
                    course=course,
                    assessment=assessment,
                    total=counseling + group + course + assessment,
                ),
            )
        )
    return ListPeopleResponse(items=items)


# ─── get_person_archive (镜像 person-archive.service.ts:253-479) ─


async def get_person_archive(
    db: AsyncSession,
    org_id: str,
    user_id: str,
) -> PersonArchive:
    """单人完整档案: 5 query 并行 (user + 4 模块 touchpoint).

    Raises:
        NotFoundError: 用户不存在 (即使有零 touchpoint, user 行存在也会成功返回
                       空 archive — 与 Node 一致)。
    """
    parse_uuid_or_raise(org_id, field="orgId")
    user_uuid = parse_uuid_or_raise(user_id, field="userId")

    # 5 个查询并行 — asyncio.gather 而非串行 await (与 Node Promise.all 等价)
    user_q = select(User).where(User.id == user_uuid).limit(1)
    episode_q = (
        select(CareEpisode)
        .where(
            and_(
                CareEpisode.org_id == parse_uuid_or_raise(org_id, field="orgId"),
                CareEpisode.client_id == user_uuid,
            )
        )
        .order_by(desc(CareEpisode.updated_at))
    )
    group_q = (
        select(GroupEnrollment, GroupInstance)
        .join(GroupInstance, GroupInstance.id == GroupEnrollment.instance_id)
        .where(
            and_(
                GroupEnrollment.user_id == user_uuid,
                GroupInstance.org_id == parse_uuid_or_raise(org_id, field="orgId"),
            )
        )
        .order_by(desc(GroupInstance.updated_at))
    )
    course_q = (
        select(CourseEnrollment, CourseInstance)
        .join(CourseInstance, CourseInstance.id == CourseEnrollment.instance_id)
        .where(
            and_(
                CourseEnrollment.user_id == user_uuid,
                CourseInstance.org_id == parse_uuid_or_raise(org_id, field="orgId"),
            )
        )
        .order_by(desc(CourseInstance.updated_at))
    )
    assessment_q = (
        select(AssessmentResult, Assessment)
        .join(Assessment, Assessment.id == AssessmentResult.assessment_id)
        .where(
            and_(
                AssessmentResult.org_id == parse_uuid_or_raise(org_id, field="orgId"),
                AssessmentResult.user_id == user_uuid,
                AssessmentResult.deleted_at.is_(None),
            )
        )
        .order_by(desc(AssessmentResult.created_at))
    )

    (
        user_result,
        episodes_result,
        groups_result,
        courses_result,
        assessments_result,
    ) = await asyncio.gather(
        db.execute(user_q),
        db.execute(episode_q),
        db.execute(group_q),
        db.execute(course_q),
        db.execute(assessment_q),
    )

    user_row = user_result.scalar_one_or_none()
    if user_row is None:
        raise NotFoundError("User", user_id)

    episodes = list(episodes_result.scalars().all())
    group_rows = list(groups_result.all())
    course_rows = list(courses_result.all())
    assessment_rows = list(assessments_result.all())

    services: list[ArchivedService] = []
    timeline: list[ArchiveTimelineEvent] = []

    # ─ counseling (镜像 service.ts:307-341) ─
    for e in episodes:
        services.append(
            ArchivedService(
                id=str(e.id),
                kind="counseling",
                org_id=str(e.org_id),
                title=user_row.name or "未知来访者",
                status=_STATUS_MAP_EPISODE.get(e.status or "", "draft"),
                description=e.chief_complaint,
                joined_at=_iso(e.opened_at),
                last_activity_at=_iso(e.updated_at) or "",
                instance_id=str(e.id),
                chief_complaint=e.chief_complaint,
                current_risk=e.current_risk,
                total_score=None,
            )
        )
        if e.opened_at:
            timeline.append(
                ArchiveTimelineEvent(
                    id=f"ep-open-{e.id}",
                    kind="counseling",
                    type="episode_opened",
                    at=_iso(e.opened_at) or "",
                    title="建立个案",
                    detail=e.chief_complaint,
                    service_id=str(e.id),
                )
            )
        if e.closed_at:
            timeline.append(
                ArchiveTimelineEvent(
                    id=f"ep-close-{e.id}",
                    kind="counseling",
                    type="episode_closed",
                    at=_iso(e.closed_at) or "",
                    title="个案结案",
                    service_id=str(e.id),
                )
            )

    # ─ group (镜像 service.ts:343-374) ─
    for grp_row in group_rows:
        enr: GroupEnrollment = grp_row[0]
        inst: GroupInstance = grp_row[1]
        joined_iso = _iso(enr.enrolled_at) or _iso(getattr(enr, "created_at", None))
        services.append(
            ArchivedService(
                id=str(inst.id),
                kind="group",
                org_id=str(inst.org_id),
                title=inst.title,
                status=_STATUS_MAP_GROUP.get(inst.status or "", "draft"),
                description=inst.description,
                joined_at=joined_iso,
                last_activity_at=_iso(inst.updated_at) or "",
                instance_id=str(inst.id),
            )
        )
        if joined_iso:
            timeline.append(
                ArchiveTimelineEvent(
                    id=f"grp-{enr.id}",
                    kind="group",
                    type="group_enrolled",
                    at=joined_iso,
                    title=f"加入团辅: {inst.title}",
                    service_id=str(inst.id),
                )
            )

    # ─ course (镜像 service.ts:376-403) ─
    for crs_row in course_rows:
        cen: CourseEnrollment = crs_row[0]
        cinst: CourseInstance = crs_row[1]
        joined_iso = _iso(cen.enrolled_at)
        services.append(
            ArchivedService(
                id=str(cinst.id),
                kind="course",
                org_id=str(cinst.org_id),
                title=cinst.title,
                status=_STATUS_MAP_COURSE.get(cinst.status or "", "draft"),
                description=cinst.description,
                joined_at=joined_iso,
                last_activity_at=_iso(cinst.updated_at) or "",
                instance_id=str(cinst.id),
            )
        )
        if joined_iso:
            timeline.append(
                ArchiveTimelineEvent(
                    id=f"crs-{cen.id}",
                    kind="course",
                    type="course_enrolled",
                    at=joined_iso,
                    title=f"加入课程: {cinst.title}",
                    service_id=str(cinst.id),
                )
            )

    # ─ assessment (镜像 service.ts:405-431) ─
    for row in assessment_rows:
        ar: AssessmentResult = row[0]
        a: Assessment = row[1]
        score: float | None = None
        if ar.total_score is not None:
            score = float(ar.total_score)
        services.append(
            ArchivedService(
                id=str(a.id),
                kind="assessment",
                org_id=str(a.org_id),
                title=a.title,
                status=_map_assessment_status(a.status, a.is_active),
                description=a.description,
                joined_at=_iso(getattr(ar, "created_at", None)),
                last_activity_at=_iso(a.updated_at) or "",
                instance_id=str(ar.id),
                total_score=score,
            )
        )
        timeline.append(
            ArchiveTimelineEvent(
                id=f"asm-{ar.id}",
                kind="assessment",
                type="assessment_taken",
                at=_iso(getattr(ar, "created_at", None)) or "",
                title=f"完成测评: {a.title}",
                detail=f"总分 {score}" if score is not None else None,
                service_id=str(a.id),
            )
        )

    # ─ Dedupe by (kind, id) — 同 user 同 assessment 多次提交可能产多 service rows
    seen: dict[str, ArchivedService] = {}
    for s in services:
        key = f"{s.kind}-{s.id}"
        prev = seen.get(key)
        if prev is None:
            seen[key] = s
            continue
        # 取更近的 joinedAt
        if s.joined_at and (not prev.joined_at or s.joined_at > prev.joined_at):
            prev.joined_at = s.joined_at
        # totalScore 优先非 None
        if s.total_score is not None and prev.total_score is None:
            prev.total_score = s.total_score
    unique_services = sorted(
        seen.values(),
        key=lambda s: s.last_activity_at,
        reverse=True,
    )

    # timeline 按 oldest → newest
    timeline.sort(key=lambda e: e.at)

    stats = PersonCounts(
        counseling=len(episodes),
        group=len(group_rows),
        course=len(course_rows),
        assessment=len(assessment_rows),
        total=len(episodes) + len(group_rows) + len(course_rows) + len(assessment_rows),
    )

    return PersonArchive(
        user=PersonArchiveUser(
            id=str(user_row.id),
            name=user_row.name,
            email=user_row.email,
            avatar_url=user_row.avatar_url,
        ),
        stats=stats,
        services=unique_services,
        timeline=timeline,
    )


__all__ = ["get_person_archive", "list_people"]


# ────────────────────────────────────────────────────────────────
# 静态 import 提示
# ────────────────────────────────────────────────────────────────

_ = ArchiveTimelineEventType
